lib/libvfio-user.c

```diff
@@ -927,126 +901,216 @@ device_reset(vfu_ctx_t *vfu_ctx, vfu_reset_type_t reason)
-    msg->out.iov.iov_base = malloc(msg->out.iov.iov_len);
-    if (msg->out.iov.iov_base == NULL) {
-        return -1;
-    }
-    dirty_pages_out = msg->out.iov.iov_base;
-    memcpy(dirty_pages_out, dirty_pages_in, sizeof(*dirty_pages_out));
-    dirty_pages_out->argsz = argsz;
+    return false;
+}
 
-    /*
-     * If the reply doesn't fit, reply with just the dirty pages header, giving
-     * the needed argsz. Typically this shouldn't happen, as the client knows
-     * the needed reply size and has already provided the correct bitmap size.
-     */
-    if (dirty_pages_in->argsz >= argsz) {
-        void *bitmap_out = msg->out.iov.iov_base + sizeof(*dirty_pages_out)
-                           + sizeof(*range_out);
-        range_out = msg->out.iov.iov_base + sizeof(*dirty_pages_out);
-        memcpy(range_out, range_in, sizeof(*range_out));
-        ret = dma_controller_dirty_page_get(vfu_ctx->dma,
-                                            (vfu_dma_addr_t)(uintptr_t)range_in->iova,
-                                            range_in->size,
-                                            range_in->bitmap.pgsize,
-                                            range_in->bitmap.size, bitmap_out);
-        if (ret != 0) {
-            ret = errno;
-            vfu_log(vfu_ctx, LOG_WARNING,
-                    "failed to get dirty bitmap from DMA controller: %m");
-            free(msg->out.iov.iov_base);
-            msg->out.iov.iov_base = NULL;
-            msg->out.iov.iov_len = 0;
-            return ERROR_INT(ret);
-        }
-    } else {
-        vfu_log(vfu_ctx, LOG_ERR,
-                "dirty pages: get [%#llx, %#llx): buffer too small (%u < %zu)",
-                (ull_t)range_in->iova, (ull_t)range_in->iova + range_in->size,
-                dirty_pages_in->argsz, argsz);
+static bool
+is_dma_feature(uint32_t feature) {
+    switch (feature) {
+        case VFIO_DEVICE_FEATURE_DMA_LOGGING_START:
+        case VFIO_DEVICE_FEATURE_DMA_LOGGING_STOP:
+        case VFIO_DEVICE_FEATURE_DMA_LOGGING_REPORT:
+            return true;
     }
 
-    return 0;
+    return false;
 }
 
 static int
-handle_dirty_pages(vfu_ctx_t *vfu_ctx, vfu_msg_t *msg)
+handle_device_feature(vfu_ctx_t *vfu_ctx, vfu_msg_t *msg)
 {
-    struct vfio_user_dirty_pages *dirty_pages = msg->in.iov.iov_base;
-    int ret;
-
     assert(vfu_ctx != NULL);
     assert(msg != NULL);
 
-    if (msg->in.iov.iov_len < sizeof(*dirty_pages) ||
-        dirty_pages->argsz < sizeof(*dirty_pages)) {
-        vfu_log(vfu_ctx, LOG_ERR, "invalid message size %zu", msg->in.iov.iov_len);
+    if (vfu_ctx->migration == NULL ||
+        msg->in.iov.iov_len < sizeof(struct vfio_user_device_feature)) {
         return ERROR_INT(EINVAL);
     }
 
-    if (vfu_ctx->migration == NULL) {
-        vfu_log(vfu_ctx, LOG_ERR, "migration not configured");
-        return ERROR_INT(ENOTSUP);
+    struct vfio_user_device_feature *req = msg->in.iov.iov_base;
+
+    uint32_t supported_flags =
```

Comment:
also rather than repeating the masking, consider:

 uint32_t operations = req->flags & ~VFIO_DEVICE_FEATURE_MASK;
 uint32_t feature = req->flags & VFIO_DEVICE_FEATURE_MASK;

before checking.